Refactor restore code and avoid rolling restart

jsanda opened the pull request:
Fixes https://github.com/k8ssandra/k8ssandra/issues/742



┆Issue is synchronized with this [Jira Bug](https://k8ssandra.atlassian.net/browse/K8SSAND-532) by [Unito](https://www.unito.io)

The initial purpose of this PR was to change the restore controller to avoid the unnecessary datacenter rolling restart. The code was a bit of a mess so I decided to take the extra time to refactor it and hopefully improve and simplify it along the way. 

Here is a brief summary of the changes:

* I added a `RestoreRequest` object that stores state for the request and also tracks changes.
* The restore start time in the status is now set immediately.
* The status now has a `DatacenterStopped` timestamp that is used how the start time was previously used. It is used to help determine whether the pod template spec updates have been applied.
* There were several `Patch` and `Update` calls scattered throughout the controller. Those have been consolidated into a single function.

The changes break the integration test; however, the restore testing was incomplete. I already have #47 in progress for replacing Ginkgo. It already has a more complete integration test for restores. I would like to keep the test changes in that PR.

I would like to have a single call to `applyUpdates()` with a `defer` statement. I need to review the code some more to see if that will work.

jsanda commented:
I think this refactoring is a big improvement. One of the things I do not like though is that I am still setting the start time at the same point, which is after the datacenter is stopped. If we change that then we may need to introduce a new field in the status to handle the logic where the start time is currently used.

burmanm commented on controllers/cassandrarestore_controller.go:
```diff
@@ -52,221 +53,171 @@ type CassandraRestoreReconciler struct {
-			return ctrl.Result{RequeueAfter: 10 * time.Second}, nil
+	if request.Restore.Spec.Shutdown && request.Restore.Status.DatacenterStopped.IsZero() {
+		if stopped := stopDatacenter(request); !stopped {
+			return r.applyUpdatesAndRequeue(ctx, request)
```
And change the restoreKey. They're not protected in any case, but each reconcile loop changes them?

jsanda replied:
`applyUpdatesAndRequeue()` checks to see if the CassandraRestore status has changed. If it has then the status is patched. It then checks to see if the CassandraDatacenter spec has changed. If it has, then the CassandraDatacenter is patched. If there are no changes, no updates should be made.

burmanm replied:
Actually only in the case that SetRestoreStartTime did not actually check that if it was not changed. I don't really like this that the process isn't visible at all, but all rely on hidden side-effects which then create more side-effects. If applyUpdates doesn't applyUpdates because SetRestoreStartTime did not do SetRestoreStartTime, then that's kinda messy approach.

jsanda replied:
>Since due to the previous "request.SetRestoreStartTime(..)" we keep updating that one with a new time all the time.
>
>And change the restoreKey. They're not protected in any case, but each reconcile loop changes them?

`SetRestoreStartTime()` and `SetRestoreStartTime()` are both idempotent. I added docs for both functions that mention this. Once they are persisted in the status, they won't get changed in subsequent reconcile loops.

>I don't really like this that the process isn't visible at all, but all rely on hidden side-effects which then create more side-effects. If applyUpdates doesn't applyUpdates because SetRestoreStartTime did not do SetRestoreStartTime, then that's kinda messy approach.

Would you find it more clear if I renamed `SetRestoreStartTime` to something like `CheckRestoreStartTime` or `SetRestoreStartTimeIfNecessary`?

I certainly could implement the logic directly in the `Reconcile` function along the following lines:

```go
if restore.Status.StartTime.IsZero() {
    restore.Status.StartTime = metav1.Now()
}

// continue with other reconciliation stuff....

if restoreStatusUpdated() {
    updateRestoreStatus()
}
```

That's where we are with the code in master and it is a mess. The additional abstraction of `SetRestoreStartTime` makes it easier to unit test that logic.

As for `applyUpdates` it simply checks to see if whether the status has changed. If it has, then it patches the status. Compared to master, I think this is a lot simpler and clearer. There is far less branching logic. Writes to the api server are limited to `applyUpdates` vs being scattered in multiple places. Lastly, the changes facilitate a better separation of concerns. The reconcile logic can focus on what changes it needs to make to the CassandraRestore and CassandraDatacenter. The `applyUpdates` function focuses on persisting those state changes.

burmanm commented:
Can you rebase the newer test so we can see this passes it?

jsanda commented:
> Can you rebase the newer test so we can see this passes it?

Sure

jsanda commented:
> Can you rebase the newer test so we can see this passes it?

jsanda commented:
@burmanm I rebased and updated the restore integration test. There are a couple things I want to note.

First, I have a big comment in the test pointing out that it does not verify the StatefulSet(s) check that the restore controller now performs. I don't like the idea of having the test go another level deeper and creating the StatefulSets that cass-operator creates. I think it gets too deep into implementation details and makes the test brittle. Open to discussion on this topic though.

Secondly, the restore controller always requeues with a delay of 10 seconds. This caused the test to take 40 sec to complete on my machine. I parameterized the delay in order to speed up the test. With the change I had the test complete in 16 sec on the last run.
